templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml

```diff
@@ -66,7 +66,19 @@ contents:
         while true; do
             if [ "$SINGLE_STACK_IPV6" = true ]
             then
-                hostname=$(curl http://[fe80::a9fe:a9fe%25br-ex]:80/openstack/2012-08-10/meta_data.json | jq -re .name)
+                metadata_ip6="fe80::a9fe:a9fe"
+                # `ip route get` returns one line and the format is:
+                # ```
+                # fe80::a9fe:a9fe dev eth0 proto kernel src fe80::f816:3eff:fe3f:da0d metric 1024 pref medium
+                # ```
+                # The device name is right after the `ðev` keyword.
+                nic=$(ip -6 route get ${metadata_ip6} | awk '/dev/{print $1}')
```

Comment:
I don't think this is right, because it only returns a single interface and there's no guarantee that it's the correct one.

For example, my system has 2 interfaces. It doesn't actually have IPv6, but both of them still publish an IPv6 link local route:

```
> ip -6 route
fc00:f853:ccd:e793::/64 dev br-ab9b6fbec819 proto kernel metric 256 linkdown pref medium
fe80::/64 dev eth0 proto kernel metric 256 pref medium
fe80::/64 dev br-ab9b6fbec819 proto kernel metric 256 linkdown pref medium
```

Unlike non-link-local addresses, `fe80::/64` on `eth0` and `br-ab9b6fbec819` are different because they are on different links. Assuming metadata was only listening on one of them, sending a metadata request to the wrong one would not work.

`ip route get` returns only a single interface, but we don't know how it chooses one over the other. It certainly cannot base the choice on the availability of the metadata service on the link. Therefore, if we use `ip route get` the correctness of this script would be based solely on factors we haven't considered, and possibly don't control.

Unless we are able to specify parameters to `ip route get` which guarantee it will always return the link with the nova metadata service, we should not use it. If we *can* work out how to do this it would be ideal, btw, but we're not that here. We must be able to guarantee, for example, that it will work correctly regardless of how many interfaces are attached, and in what order.

If we can't do that, I suggest that we just list all of them and try them in a loop, being careful to pass `--connect-timeout` of a couple of seconds at most so we don't hang on interfaces with no metadata service.

How about something like:
```bash
    ip -6 route show to match fe80::a9fe:a9fe | awk '{print $3}' | while read nic; do
        ...
        hostname=$(curl --connect-timeout 2 http://[${metadata_ip6}%25${nic}]:80/openstack/2012-08-10/meta_data.json ...)
        [ -n "$hostname" ] && break
        ...
    done
```

We only need a single pass through them all. Global retry is handled by the systemd unit.

Reply:
Thanks for your review, this was super useful. I took some of your suggestions and adjusted it a bit (with comments). Please provide feedback.